### ISSUE TITLE
assertion failed: self.suspended_since.get().is_some()

### ISSUE THREAD

**Issue by JimPanic** [servo/servo#13716]
assertion failed: self.suspended_since.get().is_some()
### Servo Version:

Servo 0.0.1-cf2a647
### Backtrace:

```

WARNING: <Constellation>: Frame FrameId { namespace_id: PipelineNamespaceId(0), index: FrameIndex(0) } iterated after closure.
WARNING: <Constellation>: Frame FrameId { namespace_id: PipelineNamespaceId(0), index: FrameIndex(0) } iterated after closure.
WARNING: <Constellation>: Frame FrameId { namespace_id: PipelineNamespaceId(0), index: FrameIndex(0) } iterated after closure.
WARNING: <ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }>: FIXME: </script> not fully implemented
WARNING: <ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }>: stop_parsing not implemented, full speed ahead!
WARNING: <Constellation>: Pipeline PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) } has no parent.
WARNING: <ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }>: stop_parsing not implemented, full speed ahead!
WARNING: <Constellation>: Frame FrameId { namespace_id: PipelineNamespaceId(1), index: FrameIndex(0) } iterated after closure.
WARNING: <ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }>: FIXME: </script> not fully implemented
WARNING: <ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }>: stop_parsing not implemented, full speed ahead!
WARNING: <ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }>: stop_parsing not implemented, full speed ahead!
WARNING: <ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(3) }>: stop_parsing not implemented, full speed ahead!
WARNING: <Constellation>: invalid traversal delta
WARNING: <Constellation>: Closing frame FrameId { namespace_id: PipelineNamespaceId(1), index: FrameIndex(4) } twice.
WARNING: <ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(5) }>: stop_parsing not implemented, full speed ahead!
WARNING: <ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(6) }>: stop_parsing not implemented, full speed ahead!
WARNING: <ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }>: Mozbrowser event after iframe PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }/PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(5) } closed.
WARNING: <ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }>: Mozbrowser event after iframe PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }/PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(5) } closed.
WARNING: <ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }>: Mozbrowser event after iframe PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }/PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(5) } closed.
WARNING: <ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }>: Mozbrowser event after iframe PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }/PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(5) } closed.
WARNING: <ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }>: Mozbrowser event after iframe PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }/PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(5) } closed.
WARNING: <ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }>: Mozbrowser event after iframe PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }/PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(5) } closed.
WARNING: <ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(5) }>: Panic hook called.
ERROR: assertion failed: self.suspended_since.get().is_some()

stack backtrace:
   0:        0x10c18535e - backtrace::backtrace::trace::h6ac07ced1b846a59
   1:        0x10c18564c - backtrace::capture::Backtrace::new::h1d3339ed08c7f861
   2:        0x10c12a645 - constellation::constellation::log_entry::hc92ce9b47dde5c21
   3:        0x10c129512 - _<constellation..constellation..FromCompositorLogger as log..Log>::log::h27ef5afa70002f0a
   4:        0x10c8abf41 - log::__log::hf7539da7e3bbacdb
   5:        0x10afd88e0 - servo::main::_{{closure}}::h70a6b1b48115735a
   6:        0x10c90a4c3 - std::panicking::rust_panic_with_hook::hcd9d05f53fa0dafc
   7:        0x10b36a204 - std::panicking::begin_panic::hc03e2830c2c89a5f
   8:        0x10b7e0be2 - script::dom::window::Window::thaw::h5f73119b8a061aaa
   9:        0x10b81f95a - script::script_thread::ScriptThread::handle_msg_from_constellation::hf1ffcabf4a69cb0b
  10:        0x10b9c6e9c - script::script_thread::ScriptThread::handle_msgs::_{{closure}}::h88b2cc62dda1cfcf
  11:        0x10b81a937 - script::script_thread::ScriptThread::handle_msgs::he9b827863d7f1f94
  12:        0x10b38cf87 - std::panicking::try::do_call::h17c16b73648098a9
  13:        0x10c90b4ea - __rust_maybe_catch_panic
  14:        0x10b560b66 - _<F as alloc..boxed..FnBox<A>>::call_box::ha39817cf138bfefc
  15:        0x10c9093f4 - std::sys::thread::Thread::new::thread_start::h50b05608a499d2b2
  16:     0x7fffc6da6aba - _pthread_body
  17:     0x7fffc6da6a06 - _pthread_start
```

This report was generated by the browser.html issue reporter.

**Comment by jdm** [comment]
A window was thawed without being previously frozen, apparently.

**Comment by jdm** [comment]
Possibly `WARNING: <Constellation>: invalid traversal delta` is related.

**Comment by ConnorGBrewster** [comment]
Is there a specific URL or test case where this can be reproduced?

**Comment by JimPanic** [comment]
Sorry for the late reply.

This happened in the following stream of events IIRC:

- open browser
- enter some URL
- enter different (internal) URL because I'm behind a corporate proxy and there's no proxy support :)
- hit back
-> error

**Comment by emilio** [comment]
I'm leaving this as the canonical issue for this panic. https://github.com/servo/servo/issues/13780 also contains relevant information.

**Comment by emilio** [comment]
Also, given the amount of duplicates, people seem to hit this a lot. I don't have the time right now to investigate it, but it'd be nice if someone could take a look at it.

**Comment by cbrewster** [comment]
I'll go ahead and take a look at this one since it seems traversal related.

**Comment by cbrewster** [comment]
So there seems to be a case where a pipeline does not get frozen and on history traversal, we attempt to thaw the pipeline which causes the assertion because the pipeline was never frozen.

**Comment by cbrewster** [comment]
cc @asajeffrey

**Comment by cbrewster** [comment]
When `bind_to_tree` gets called we sometimes will call:
```
self.create_nested_browsing_context();
self.process_the_iframe_attributes(ProcessingMode::FirstTime);
```
https://dxr.mozilla.org/servo/source/components/script/dom/htmliframeelement.rs#702-703
Both `self.create_nested_browsing_context` and `self.process_the_iframe_attributes` call `navigate_or_reload_child_browsing_context`. In the case that both of those have the `about:blank` url, then the `ScriptLoadedAboutBlankInIFrame` is sent twice to the constellation. This initiates 2 FrameChanges on the same Frame with 2 new Pipelines. When we navigate/traverse only one of those will be frozen while the other will remain unfrozen. When we try to thaw the one that was not frozen during traversal, we will encounter the panic.

**Comment by cbrewster** [comment]
Here is a naive solution; however, I think there is some cleanup to be done here.
```
---
 components/script/dom/htmliframeelement.rs | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/components/script/dom/htmliframeelement.rs b/components/script/dom/htmliframeelement.rs
index 53c7e34..b15976c 100644
--- a/components/script/dom/htmliframeelement.rs
+++ b/components/script/dom/htmliframeelement.rs
@@ -699,7 +699,9 @@ impl VirtualMethods for HTMLIFrameElement {
         if self.upcast::<Node>().is_in_doc_with_browsing_context() {
             debug!("iframe {} bound to browsing context.", self.frame_id);
             debug_assert!(tree_in_doc, "is_in_doc_with_bc, but not tree_in_doc");
-            self.create_nested_browsing_context();
+            if !self.upcast::<Element>().has_attribute(&local_name!("src")) {
+                self.create_nested_browsing_context();
+            }
             self.process_the_iframe_attributes(ProcessingMode::FirstTime);
         }
     }
--
2.9.2
```

**Comment by asajeffrey** [comment]
I'd be quite keen on making freeze and thaw idempotent, which would fix this. We should try to fix the underlying problem though, we shouldn't have multiple pipelines for the same about:blank.

**Comment by jdm** [comment]
I don't believe the patch in the previous comment is a correct solution, since the nested browsing context would no longer be synchronously available for `<iframe></iframe>`.

**Comment by cbrewster** [comment]
This is fixed by #14312 (it makes freeze/thaw idempotent). There is still an underlying issue here, but I opened for some more specifics #14856.
